### PR TITLE
Test and implement autocomplete API

### PR DESCRIPTION
**Pull request by mdittmer**
@Hexcles this is now ready for review.

### CODE REVIEW

**Comment by Hexcles** [comment]
Sorry, this is a bit hard to review as it is right now. I'd like to wait until #414 lands.

If you would like me to take a quick look ASAP before you add more stuff, perhaps you can try to change the base of this PR to the feature branch of #414, which should clean up the diff quite a lot.

**Comment by mdittmer** [comment]
Don't worry about it. Let's focus on landing #414, then we can update this and review it.

**Comment by mdittmer** [comment]
Yes.

On Aug. 18, 2018 2:23 p.m., "Robert Ma" <notifications@github.com> wrote:

#414 <https://github.com/web-platform-tests/wpt.fyi/pull/414> has landed.
Is this ready for review?

(That said, I might have to delay the review a bit until I fix the tests.
Turns out in addition to integration tests, our Go lint checks are also
kind of broken...)

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/web-platform-tests/wpt.fyi/pull/436#issuecomment-414077339>,
or mute the thread

File: api/query/autocomplete.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
+// Copyright 2018 The WPT Dashboard Project. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package query
+
+import (
+	"encoding/json"
+	http "net/http"
```
Comment (Hexcles) [inline api/query/autocomplete.go]: Nit: this should simply be `"net/http"`

File: api/query/query.go [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+	"encoding/json"
+	"errors"
+	"fmt"
+	io "io"
+	"io/ioutil"
+	"log"
+	http "net/http"
+	"strconv"
+	"sync"
+	time "time"
```
Comment (Hexcles) [inline api/query/query.go]: Here are some unnecessary import renamings as well. My local `gofmt` would get rid of them.

File: api/query/autocomplete.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	// Suggestions is the collection of autocomplete suggestions.
+	Suggestions []AutocompleteResult `json:"results"`
+}
+
+type byQueryIndex struct {
+	q  string
+	rs []AutocompleteResult
+}
+
+func (r byQueryIndex) Len() int      { return len(r.rs) }
```
Comment (Hexcles) [inline api/query/autocomplete.go]: Can we document and test the sorting criteria?

**Comment by Hexcles** [comment]
I reviewed `autocomplete.go` and its tests. The code generally LGTM except some nits I commented above. Please rebase the branch and resolve the conflict. And let's see how Travis looks like after the push.

As discussed offline, the refactor of `search.go` -> `query.go` will go into another PR. And it should be a pure refactor that I will probably just rubber-stamp, so I didn't look at those changes here.

**Comment by Hexcles** [comment]
Thanks for splitting the PR. #467 was just merged. Let's rebase this one and I don't expect I'd have other significant comments besides the request for testing sorting `byQueryIndex`,

**Comment by mdittmer** [comment]
Merge seemed to go well. `byQueryIndex` [now tested](https://github.com/web-platform-tests/wpt.fyi/pull/436/files#diff-7a30f0904892d1fa90ba3ee825e5edc9R18).

**Pull request merged**